### DISCUSSION TITLE
data in pyMOR

### DISCUSSION THREAD

**Discussion by artpelling** [pymor/pymor#1644] [Ideas]
As you probably know, I would like to get more data-driven methods into pyMOR. I feel that while `VectorArrays` are great for working with `Operators` and external solvers, they add a substantial amount of (syntactical) overhead and are not particularly well suited for representing "data".

Following the discussion in #390, I spent some time thinking about what "data" actually means and came to the conclusion that it can probably be represented by a set of pairings `(x, y)` in most cases. So my idea here is to first derive some classes that represent series of `x` and `y` individually, which I called `Samples` and then have `Series` that represent pairings of the former. I am happy to change the names for something better.

I have only looked at this from the system theoretic perspective. Specifically, I have QuadBT, Loewner-based methods and AAA interpolation in mind. I guess it could be good to maybe include some notion of labeled data for AI/ML approaches...
My aim is to only have the data in memory once (in the `Samples` class) or if the data gets too large, fetch it from somewhere when needed. It could be read from a file, queried from a database or computed. The classes for this would have to be added in the future, I have implemented a `TransferFunctionSampler` as an example for on demand computation. Using the `__call__` and `__iter__` methods, quantities such as the Loewner or Hankel matrix could be then directly constructed from the data in a pipeline without further memory usage.

I am excited to hear what you think! Mind you, this is only a very basic skeleton that is still quite inconsistent, especially the inheritance is not right. I thought it would be good to check back with you all before investing too much time :)

```python
import numpy as np
import numpy.ma as ma

from pymor.core.base import BasicObject
from pymor.models.transfer_function import TransferFunction


class GenericSamples(BasicObject):
    def __init__(self, data, mask=None):
        assert hasattr(data, '__iter__')
        self.n = len(data)
        self.__auto_init(locals())

    def __call__(self, i):
        return self.data[i]

    def __iter__(self):
        for i in range(self.n):
            if not ma.is_masked(self(i)):
                yield self(i)


class NumpySamples(GenericSamples):
    def __init__(self, data, mask=ma.nomask):
        self.data = np.asarray(data) if mask is ma.nomask else ma.array(data, mask=mask)
        self.n = data.shape[0]


class TransferFunctionSampler(GenericSamples):
    def __init__(self, w, tf):
        assert isinstance(w, GenericSamples)
        assert isinstance(tf, TransferFunction)
        self.n = w.n
        self.__auto_init(locals())

    def __call__(self, i):
        return ma.masked if ma.is_masked(self.w(i)) else self.tf.eval_tf(self.w(i))


class TimeSeries(GenericSamples):
    def __init__(self, x, y):
        assert isinstance(x, GenericSamples)
        assert isinstance(y, GenericSamples)
        assert x.n == y.n
        self.n = x.n
        self.__auto_init(locals())

    def __call__(self, i):
        return ma.masked if (ma.is_masked(self.x(i)) or ma.is_masked(self.y(i))) else (self.x(i), self.y(i))


class FrequencySeries(TimeSeries):
    def __init__(self, *args):
        super().__init__(*args)

    @classmethod
    def from_transfer_function(cls, w, tf):
        assert isinstance(w, GenericSamples)
        assert isinstance(tf, TransferFunction)
        return cls(w, TransferFunctionSampler(w, tf))


if __name__ == '__main__':
    from pymor.models.iosys import LTIModel

    sys = LTIModel.from_matrices(np.random.rand(10, 10), np.ones([10, 2]), np.ones([1, 10]))
    t = NumpySamples(np.geomspace(10, 1000, 15), mask=np.random.choice([True, False], 15))
    F = FrequencySeries.from_transfer_function(t, sys.transfer_function)

    for f in F:
        print(f)
```

**Comment by renefritze** [comment]
Hey @artpelling,

I think it's great that you try to explore if we have a real need for something else besides `VectorArrays`.

Can you say more on why you feel the need for a dedicated class hierarchy for this? Why are `VectorArrays` not well suited for data (other than having functions you wouldn't need for your use-case)?

Would this carry the same abstraction level as `VectorArrays`? As in have implementations for different backends? 

The masking seems central to your use case, could this be integrated with `VectorArrays` instead?

**Reply by artpelling** [reply]
> Can you say more on why you feel the need for a dedicated class hierarchy for this? Why are `VectorArrays` not well suited for data (other than having functions you wouldn't need for your use-case)?

My concerns are that the data should be in memory only once, but this part would work also with `NumpyVectorArrays`. I don't see any occasion where one would need to use the `VectorSpaces` related structure for data representation and I think that it would be nice to have something on a lower level that is easy to work with from a user perspective.
Independently of the implementation, I think dedicated data objects for time and frequency domain data are needed that handle the mapping of time/frequency to the sample values. My aim is to
- not have `VectorSpaces`
- be able to have matrix valued samples
- have (x,y) pairs of data
- provide an easy interface

and I did not see a viable way to do this by subclassing `VectorArray`.

> Would this carry the same abstraction level as `VectorArrays`? As in have implementations for different backends?

Yes that would be the ultimate goal. I envision something that can fetch from databases, compute data on-demand and read different data formats such as HDF5 for example (ideally only the needed parts of the file). I tried to sketch the idea of on-demand computing with the `TransferFunctionSampler` with something like this, data-driven methods could be applied to `Model`s easily.

> The masking seems central to your use case, could this be integrated with `VectorArrays` instead?

The masking at the moment is heavily motivated by Loewner data partitioning and AAA data subselection. I would like to leave the data untouched in memory but create new objects of subselected data without copying. I am sure there is also a way to get this behaviour with `VectorArray`s.

**Comment by sdrave** [comment]
Hi @artpelling, while I am generally open to the idea, I am not quite sure about where this is headed. In your code above, the interface of `GenericSamples` is just that of a `Sequence`, and `TimeSeries` is just a sequence of pairs. So I do not see a real benefit in formally introducing these classes at the moment. Also, I am somewhat concerned about possible fragmentation between code which uses `VectorArrays` and code which uses `Data`. One would need to make it clear in what ways these concepts are compatible.

I have added the topic to the agenda of today's dev meeting. Maybe we can discuss then in a little bit more detail.

**Reply by artpelling** [reply]
> Hi @artpelling, while I am generally open to the idea, I am not quite sure about where this is headed. In your code above, the interface of `GenericSamples` is just that of a `Sequence`, and `TimeSeries` is just a sequence of pairs. So I do not see a real benefit in formally introducing these classes at the moment. 

You are right. This code above was only meant to serve as some basis for discussion. As per usual, after posting I have found several issues with it myself and am working on updating this to be more sensible. Currently, I am thinking about starting with sequences of pairs as a base class directly, this would also make masking way easier. Unfortunately, I did not arrive on something that is consistent and viable yet. I hope to be able to show something more refined soon.

> Also, I am somewhat concerned about possible fragmentation between code which uses `VectorArrays` and code which uses `Data`. One would need to make it clear in what ways these concepts are compatible.

I hear you. I try to view both from a mathematical standpoint - a set of vectors on one hand and a sequence of tuples $i\rightarrow (x_i, y_i)$, $x_i\in\mathbb{C}$, $y_i\in\mathbb{C}^{n\times m}$ on the other hand. In this view, it seems natural that the $y_i$ can be the elements of a `VectorArray`.

> I have added the topic to the agenda of today's dev meeting. Maybe we can discuss then in a little bit more detail.

Thanks, that would be great.
